src/flp/types/l1boundsum.rs

```diff
@@ -0,0 +1,402 @@
+
+        // Check if the bit width is too large. This limit is defined to be one bit less than the
+        // number of bits required to encode `F::Integer`. (One less so that we can compute `1 <<
+        // bits` without overflowing.)
+        let limit = std::mem::size_of::<F::Integer>() * 8 - 1;
+        if bits > limit {
+            return Err(FlpError::InvalidParameter(format!(
+                "bit width exceeds limit of {limit}"
+            )));
+        }
+
+        // Check for degenerate parameters.
+        if bits == 0 {
+            return Err(FlpError::InvalidParameter(
+                "bits cannot be zero".to_string(),
+            ));
+        }
+        if measurement_len == 0 {
+            return Err(FlpError::InvalidParameter(
+                "measurement_len cannot be zero".to_string(),
+            ));
+        }
+        if chunk_length == 0 {
+            return Err(FlpError::InvalidParameter(
+                "chunk_length cannot be zero".to_string(),
+            ));
+        }
+
+        let mut gadget_calls = measurement_len_in_bits / chunk_length;
+        if measurement_len_in_bits % chunk_length != 0 {
+            gadget_calls += 1;
+        }
+
+        Ok(Self {
+            measurement_len,
+            bits,
+            measurement_len_in_bits,
+            chunk_length,
+            gadget_calls,
+            phantom: PhantomData,
+        })
+    }
+
+    /// Computes the largest field element for the chosen bit count.
+    fn largest_field_element(&self) -> F::Integer {
+        let one = F::Integer::from(F::one());
+        (one << self.bits) - one
+    }
+}
+
+impl<F: NttFriendlyFieldElement, S> Clone for L1BoundSum<F, S> {
+    fn clone(&self) -> Self {
+        Self {
+            measurement_len: self.measurement_len,
+            bits: self.bits,
+            measurement_len_in_bits: self.measurement_len_in_bits,
+            chunk_length: self.chunk_length,
+            gadget_calls: self.gadget_calls,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<F, S> Flp for L1BoundSum<F, S>
+where
+    F: NttFriendlyFieldElement,
+    S: ParallelSumGadget<F, Mul<F>> + Eq + 'static,
+{
+    type Field = F;
+
+    fn gadget(&self) -> Vec<Box<dyn Gadget<F>>> {
+        vec![Box::new(S::new(
+            Mul::new(self.gadget_calls),
+            self.chunk_length,
+        ))]
+    }
+
+    fn num_gadgets(&self) -> usize {
+        1
+    }
+
+    fn valid(
+        &self,
+        gadget: &mut Vec<Box<dyn Gadget<F>>>,
+        input: &[F],
+        joint_randomness: &[F],
+        num_shares: usize,
+    ) -> Result<Vec<F>, FlpError> {
+        self.valid_call_check(input, joint_randomness)?;
+
+        let range_check = parallel_sum_range_checks(
+            &mut gadget[0],
+            input,
+            joint_randomness,
+            self.chunk_length,
+            num_shares,
+        )?;
+
+        // The encoded input consists of the measurement with a single field element appended,
```

Comment:
This comment is a bit misleading, because there will be `bits` many field elements appended. Rather, the claimed weight is also encoded with one field element per bit, like the numbers in the original measurement.